go/vt/tabletserver/sysloglogger/sysloglogger_test.go

```diff
@@ -0,0 +1,161 @@
+package sysloglogger
+
+import (
+	"log/syslog"
+	"testing"
+
+	"context"
+
+	"fmt"
+	"time"
+
+	"strconv"
+
+	"errors"
+	"strings"
+
+	"github.com/youtube/vitess/go/vt/tabletserver"
+)
+
+// fakeWriter is a mock, implementing the syslogWriter interface, which can be used in place of
+// the real syslog writer to enable capturing and playing back of log messages in unit testing.
+type fakeWriter struct {
+	messages map[string]bool
+	err      error // if non-nil, force an error to be returned
+}
+
+func (fw *fakeWriter) write(pri syslog.Priority, msg string) error {
+	if fw.err != nil {
+		return fw.err
+	}
+	fw.messages[strings.TrimSpace(msg)] = true
+	return nil
+}
+
+func (fw *fakeWriter) Info(msg string) error { return fw.write(syslog.LOG_INFO, msg) }
+func (fw *fakeWriter) Close() error          { return nil }
+func (fw *fakeWriter) messageCount() int     { return len(fw.messages) }
+func (fw *fakeWriter) containsMessage(msg string) bool {
+	_, ok := fw.messages[strings.TrimSpace(msg)]
+	return ok
+}
+
+// mockLogStats generates a dummy tabletserver.LogStats message for testing.
+func mockLogStats(originalSQL string) *tabletserver.LogStats {
+	logstats := tabletserver.NewLogStats("Execute", context.Background())
+	logstats.StartTime = time.Time{}
+	logstats.PlanType = "PASS_SELECT"
+	logstats.OriginalSQL = originalSQL
+	return logstats
+}
+
+// expectedLogStatsText returns the results expected from the plugin processing a dummy message generated by mockLogStats(...).
+func expectedLogStatsText(originalSQL string) string {
+	return fmt.Sprintf("Execute\t\t\t''\t''\tJan  1 00:00:00.000000\tJan  1 00:00:00.000000\t0.000000\tPASS_SELECT\t"+
+		"\"%s\"\tnull\t0\t\"\"\tnone\t0.000000\t0.000000\t0\t0\t\"\"", originalSQL)
+}
+
+// TestSyslog sends a stream of five query records to the plugin, and verifies that they are logged.
+func TestSyslog(t *testing.T) {
+	// Overwrite the usual syslog writer and StatsLogger subscription channel with mocks
+	mock := &fakeWriter{
+		messages: make(map[string]bool),
+	}
+	writer = mock
+	ch = make(chan interface{}, 10)
+
+	// Start running the plugin loop
+	go run()
+
+	// Send fake messages to the mock channel, close it to end the plugin loop, and briefly wait for things to settle
+	for i := 1; i <= 5; i++ {
+		ch <- mockLogStats("select " + strconv.Itoa(i))
+	}
+	close(ch)
+	time.Sleep(500 * time.Millisecond)
```

Comment:
So here you just need to wait until the 'run' method exits. So instead of 'go run()' above, do this:

syncChannel := make(chan bool)
go func() {
  run()
  close(syncChannel)
}()

And then here, instead of the sleep(), just do:
<-syncChannel
that will return when the channel is closed, meaning run() is done.

Using channels like this to synchronize unit tests is the best way to avoid sleeps. You can probably do similar things below